core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/AbstractAllowanceWithRanges.java

```diff
@@ -324,9 +324,17 @@ public Envelope computeIteration(Envelope base,
         var builder = new EnvelopeBuilder();
         if (leftPart != null)
             builder.addPart(leftPart);
-        builder.addParts(coreEnvelope.slice(leftPartEndPos, rightPartBeginPos));
-        if (rightPart != null)
+        if (rightPart != null) {
+            // Forcing the right part begin speed avoids an epsilon difference that would cause errors later on
+            assert abs(coreEnvelope.interpolateSpeed(rightPartBeginPos) - rightPart.getBeginSpeed()) < 1e-8;
+            builder.addParts(coreEnvelope.slice(
+                    leftPartEndPos, Double.NaN,
+                    rightPartBeginPos, rightPart.getBeginSpeed()
+            ));
```

Comment:
Nice catch! I believe the same issue can occur with the left part.
I would suggest adding two more local variables, leftPartEndSpeed and rightPartBeginSpeed, which can be set along with leftPartEndPos and rightPartBeginPos.
This way, there is a single unconditional coreEnvelope.slice call.